### PR TITLE
Fix Namespace Termination

### PR DESCRIPTION
**Pull request by derekwaynecarr**
There was a regression as part of the PrepareForUpdate changes that prevented deletion of a namespace because the list of finalizers was never being reduced because we were always taking the input from the client and overriding it with the old value from the server.

This change fixes the following:
1. Fixes issue where we copied old objectMeta to new object which made client specified resource versions ignored for idempotent calls.  This caused race conditions between multiple finalizers.
2. Add prepare for update strategy for namespace finalize strategy to do the right thing

/cc @smarterclayton

### CODE REVIEW

**Comment by vmarmol** [comment]
Assigning @smarterclayton, feel free to reassign.

**Comment by derekwaynecarr** [comment]
@smarterclayton - i will ping to review this PR, want to add a little more testing.

**Comment by smarterclayton** [comment]
K, was about to ask for that

**Comment by derekwaynecarr** [comment]
@smarterclayton - this should be good to go now functionally, i added an extra validation test with mutiple finalizers... if we want to unblock the bug, we can merge this now, and i will follow-up tomorrow with more testing to ensure no future regression.

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -1125,12 +1125,9 @@ func validateFinalizerName(stringValue string) errs.ValidationErrorList {
...
 	allErrs := errs.ValidationErrorList{}
 	allErrs = append(allErrs, ValidateObjectMetaUpdate(&oldNamespace.ObjectMeta, &namespace.ObjectMeta).Prefix("metadata")...)
-
-	// TODO: move reset function to its own location
-	// Ignore metadata changes now that they have been tested
-	oldNamespace.ObjectMeta = namespace.ObjectMeta
 	oldNamespace.Spec.Finalizers = namespace.Spec.Finalizers
-
+	oldNamespace.Labels = namespace.Labels
+	oldNamespace.Annotations = namespace.Annotations
```
Comment (smarterclayton) [inline pkg/api/validation/validation.go]: Why this change?

Comment (derekwaynecarr) [inline pkg/api/validation/validation.go]: this is what happens when you are rushing to fix something before leaving work ;-)

cleaning this up more this morning and getting more tests for various strategies.

**Comment by derekwaynecarr** [comment]
@smarterclayton - added tests for each of the strategies that they are doing the right thing, and fixed where we had flipped old and new on validateNamespaceUpdate call.

**Comment by smarterclayton** [comment]
LGTM
